filter.c

```diff
@@ -1168,6 +1176,135 @@ static int func_avg(filter_t *flt, bcf1_t *line, token_t *rtok, token_t **stack,
     }
     return 1;
 }
+static int compare_doubles(const void *lhs, const void *rhs)
+{
+    double arg1 = *(const double*) lhs;
+    double arg2 = *(const double*) rhs;
+    if (arg1 < arg2) return -1;
+    if (arg1 > arg2) return 1;
+    return 0;
+}
+static double median(double *vals, const int n)
+{
+    switch (n) {
+        case 0: return 0;
+        case 1: return vals[0];
+        case 2: return (vals[0] + vals[1]) / 2;
+        default:
+        {
+            qsort(vals, n, sizeof(double), compare_doubles);
+            const int middle = n / 2;
+            if (n % 2 == 1) {
+                return vals[middle];
+            } else {
+                return (vals[middle - 1] + vals[middle]) / 2;
+            }
+        }
+    }
+}
+static int func_median(filter_t *flt, bcf1_t *line, token_t *rtok, token_t **stack, int nstack)
+{
+    token_t *tok = stack[nstack - 1];
+    rtok->nvalues = 0;
+    if ( !tok->nvalues ) return 1;
+    double *vals = (double*) malloc(tok->nvalues * sizeof(double));
+    int i, n = 0;
+    for (i=0; i<tok->nvalues; i++)
+        if ( !bcf_double_is_missing(tok->values[i]) ) { vals[n] = tok->values[i]; n++; }
+    if ( n )
+    {
+        rtok->values[0] = median(vals, n);
+        rtok->nvalues   = 1;
+    }
+    free(vals);
+    return 1;
+}
+static double mode(double *vals, const int n)
+{
+    switch (n) {
+        case 0: return 0;
+        case 1: return vals[0];
+        case 2: return vals[0] < vals[1] ? vals[0] : vals[1];
+        default:
+        {
+            qsort(vals, n, sizeof(double), compare_doubles);
+            // discretize real values
+            int *int_vals = (int*) malloc(n * sizeof(int));
+            int i;
+            for (i=0; i<n; i++) int_vals[i] = vals[i];
+            // find mode of discrete samples
+            int mode_count = 0, mode_idx = 0;
+            for (i=0; i<n; i++) {
+                int j;
+                for (j = i + 1; j < n && vals[i] == vals[j]; ++j);
+                int count = j - i;
+                if (count > mode_count) {
+                    mode_idx = i;
+                    mode_count = count;
+                }
+            }
+            free(int_vals);
+            return median(vals + mode_idx, mode_count);
+        }
+    }
+}
+static int func_mode(filter_t *flt, bcf1_t *line, token_t *rtok, token_t **stack, int nstack)
```

Comment:
`func_mode`, `func_median` and `func_stdev` are similar. It would be preferable to have a single method with an additional parameter in the form of a function pointer, which, in the actual call, will be replaced by `mode`, `median` or `stdev`.

Reply:
Addressed in 9d657a5edf2e71b3aa98fe9ac52ff17915af049e.